hishel/_utils.py

```diff
@@ -33,17 +33,36 @@ def normalized_url(url: tp.Union[httpcore.URL, str, bytes]) -> str:
     assert False, "Invalid type for `normalized_url`"  # pragma: no cover
 
 
-def generate_key(request: httpcore.Request) -> str:
+def sync_generate_body_hash(request: httpcore.Request) -> str:
+    if isinstance(request.stream, tp.Iterable):
+        body_hash = blake2b(digest_size=16)
+        for chunk in request.stream:
+            body_hash.update(chunk)
+    elif isinstance(request.stream, tp.AsyncIterable):
+        raise TypeError("AsyncIterable streams are not supported for request bodies in the sync client.")
+    return body_hash.hexdigest()
+
+
+async def async_generate_body_hash(request: httpcore.Request) -> str:
+    if isinstance(request.stream, tp.AsyncIterable):
+        body_hash = blake2b(digest_size=16)
+        async for chunk in request.stream:
+            body_hash.update(chunk)
+    elif isinstance(request.stream, tp.Iterable):
+        body_hash = blake2b(digest_size=16)
+        for chunk in request.stream:
+            body_hash.update(chunk)
+    return body_hash.hexdigest()
+
+
+def generate_key(request: httpcore.Request, body_hash: tp.Optional[str] = None) -> str:
     encoded_url = normalized_url(request.url).encode("ascii")
-
-    key_parts = [
-        request.method,
-        encoded_url,
-    ]
-
     key = blake2b(digest_size=16)
-    for part in key_parts:
-        key.update(part)
+    key.update(request.method)
+    key.update(encoded_url)
+    if body_hash is None and isinstance(request.stream, (tp.Iterable, bytes)):
```

Comment:
The thinking here is that if `body_hash` is None (i.e. for some reason no specific body hash is provided), then the `generate_key` function examines `httpcore.Request.stream` to see if it has a body that can be hashed synchronously (i.e. is of type `Iterable|bytes`), and if so, it hashes it. 

Maybe the more elegant way of handling it is to make `body_hash` required (type `str`), since the only places it is called will definitely provide that parameter, and then removing that code.

Also - I think there is another error here. If this code isn't called, the hash should still be updated with the body_hash value. So `key.update(body_hash.encode("ascii")` should be outside the `if` body. 

I'll push an update this afternoon to fix that (or confirm that this GitHub display is incorrect).

Thanks!

Reply:
I've made a change to account for this.

I've left the default `None` for body hash, and kept the synchronous hashing logic just so that custom "cache_key" functions can still work even if they don't use the `body_hash` parameter. It also keeps existing tests working that don't expressly pass a `body_hash` parameter.

After that logic, and assuming that the body hash is not none, the body hash is included in the cache key hash.

Reply:
> and kept the synchronous hashing logic just so that custom "cache_key" functions can still work even if they don't use the body_hash parameter

I don't think I understand synchronous hashing logic help custom function like:

```python
def custom_key_generator(request: httpcore.Request, body_hash = None):
    key = generate_key(request)
    method = request.method.decode()
    host = request.url.host.decode()
    return f"{method}|{host}|{key}"
```

How we want this to work in async|sync case?
Is there a place when we want to call generate_key without body_hash (except for the tests)?